### ISSUE TITLE
Support Packer

### ISSUE THREAD

**Issue by rarkins** [renovatebot/renovate#8178]
**What would you like Renovate to be able to do?**

Keep AMI image references up-to-date using Renovate

**Did you already have any implementation ideas?**

Would need a new datasource as well as manager. Also I'm not sure they use *versions* so we may need a new concept such as sorting versions by publish date.

Community input on how this could/should work would be appreciated.

**Comment by secustor** [comment]
I guess you mean to update the source images in packer definitions? There is also the usecase to update these images in Terraform code.

Packer itself has no concept of versions.   
The most used method is to select a "image stream" based on filters and properties e.g. the owner and then use the newest.     
To provide updates we would need to reflect these in Renovate, Package Rules maybe?     
The same logic is used to select vm images in Terraform. 
An alternativ could be to tag the images with versions or build numbers, but this can not be expected that every organization follows this pattern. 

We probably also need for every image provider a own datasource as they have different image properties and API schemes. Further we have to think about authentification as some cloud providers require a login to query plublic images as well as private ones. 

This is maybe already solved, but it needs to be clarified how to inject the cloud provider credentials into the runtime of Renovate. There are ways around in the self hosted versions, e.g. a own renovate instance for the project or a cloud provider user to list all images inside an organization. 
These workarounds are not really viable from my point of view for the hosted versions of Renovate.

The login mechanism could also be used in Terraform to implement updates of Resources in Renovate e.g. upgrade the Kubernetes version of a deployed cluster.   
This would be another step for Renovate to be the goto Updater for IaC. 

https://www.orsolabs.com/post/packer-terraform-images-versioned-dry/

These are just my initial thoughts on a Packer implementation in Renovate :)

**Comment by rarkins** [comment]
> The most used method is to select a "image stream" based on filters and properties e.g. the owner and then use the newest.
> To provide updates we would need to reflect these in Renovate, Package Rules maybe?

Yes, I was thinking about this as a way to "pin" AMIs but stay updated instead of using non-deterministic filters in terraform/packer definitions. Possibly we could reuse the `followTag` concept from npm and "follow filter"?

> This is maybe already solved, but it needs to be clarified how to inject the cloud provider credentials into the runtime of Renovate.

We do this reasonably well with `hostRules` already although we maybe need to add some adapters for cloud providers which require a *dynamic* token be generated from other credentials.

**Comment by secustor** [comment]
I couldn't find any usage examples for `followTag`. Do you have some examples how it works? 

Wouldn't `hostRules` require the user to know the URL for the API for the CloudProvider/ImageSource instead of simply choose the provider e.g. Amazon and then put in their configurations?    
Further I think we should not propagate to put secrets unencrypted inside the Git repo. Maybe we could use something like [Ansible Vault](https://docs.ansible.com/ansible/latest/user_guide/vault.html).

**Comment by mvanholsteijn** [comment]
For inspiration: I actually created an packer image update utility for Google VM images. https://binx.io/blog/2021/04/11/how-to-keep-your-google-packer-source-image-version-up-to-date/

Getting credentials to access the API is a thing, indeed.

**Comment by ibacalu** [comment]
This should work out-of-the-box, but needs testing:

### Renovate config
```json
{
    "regexManagers": [
        {
            "fileMatch": [
                "\\.pkr\\.hcl"
            ],
            "matchStrings": [
                "renovate: (\\s*sourceUrl=(?<sourceUrl>[^\\s]+)){0,1}(\\s*datasource=(?<datasource>[^\\s]+)){0,1}(\\s*depName=(?<depName>[^\\s]+))*\\n.*= \"(?<currentValue>[^\\s]+)\""
            ]
        }
    ]
}
```
and below is the packer structure (or whatever file defined in `fileMatch`)
```hcl
# somefile.pkr.hcl
variable "image" {
  type    = string
  # renovate: sourceUrl=https://github.com/repo/ami-build datasource=github-releases depName=ami-build
  default = "ami-0123"
}
```
This way the `datasource`, `depname` and `sourceUrl` are defined in the comment inside the file and you can thus have multiple different `datasources` in the same document.

**Comment by Fredouye** [comment]
Thanks @ibacalu for the syntax.

It's working for me (with Renovate 36.89.0) and this configuration :

- `renovate.json`

```
{
  "regexManagers": [
    {
        "fileMatch": [
            "\\.pkr\\.hcl"
        ],
        "matchStrings": [
                "renovate: (\\s*sourceUrl=(?<sourceUrl>[^\\s]+)){0,1}(\\s*datasource=(?<datasource>[^\\s]+)){0,1}(\\s*depName=(?<depName>[^\\s]+))*\\n.*= \"(?<currentValue>[^\\s]+)\""
        ],
        "datasourceTemplate": "github-releases"
    }
  ]
}
```

- `<subdir>/plugins.pkr.hcl`

```
packer {
  required_plugins {
    vsphere = {
      # renovate: depName=hashicorp/packer-plugin-vsphere
      version = "1.2.0"
      source = "github.com/hashicorp/vsphere"
    }
  }
}
```

**Comment by M0NsTeRRR** [comment]
I'm using an easier solution that doesn't need to set depName, you can set version after source or vice versa it doesn't matter and it also catch packer `required_version`

regexManagers
```
  {
    fileMatch: [
      '\\.pkr\\.hcl',
    ],
    matchStrings: [
      'required_version[\\s]+=[\\s]+"=?(?<currentValue>\\S*)"',
      'source[\\s]+=[\\s]+"github.com/(?<depName>\\S*)"[\\s]+version[\\s]+=[\\s]+"(?<currentValue>\\S*)"',
      'version[\\s]+=[\\s]+"(?<currentValue>\\S*)[\\s]+source[\\s]+=[\\s]+"github.com/(?<depName>\\S*)"',
    ],
    datasourceTemplate: 'github-tags',
    depNameTemplate: '{{#if depName}}{{#if (containsString depName "hashicorp/")}}{{{replace "hashicorp/" "hashicorp/packer-plugin-" depName}}}{{else}}{{{depName}}}{{/if}}{{else}}hashicorp/packer{{/if}}',
  }
```

packer
```
packer {
  required_version = "v1.9.4"
  required_plugins {
    proxmox = {
      source  = "github.com/hashicorp/proxmox"
      version = "1.1.3"
    }
    ansible = {
      source  = "github.com/hashicorp/ansible"
      version = "1.1.0"
    }
  }
}
```

**Comment by Fredouye** [comment]
Thanks @M0NsTeRRR, it's indeed easier :)

**Comment by MPV** [comment]
Should this issue be split into:
1. Renovate to support upgrading Packer plugins
2. Renovate to support upgrading Packer source images (AMIs etc)
